gtsfm/frontend/correspondence_generator/mast3r_correspondence_generator.py

```diff
@@ -0,0 +1,341 @@
+        self,
+        client: Client,
+        images: List[Future],
+        image_pairs: List[Tuple[int, int]],
+    ) -> Tuple[List[Keypoints], Dict[Tuple[int, int], np.ndarray]]:
+        """Apply the correspondence generator to generate putative correspondences.
+
+        Args:
+            client: Dask client, used to execute the front-end as futures.
+            images: List of all images, as futures.
+            image_pairs: Indices of the pairs of images to estimate two-view pose and correspondences.
+
+        Returns:
+            List of keypoints, one entry for each input images.
+            Putative correspondence as indices of keypoints, for pairs of images.
+        """
+        device = torch.device("cuda" if torch.cuda.is_available() else "cpu")
+        model = AsymmetricMASt3R.from_pretrained(_MODEL_PATH).eval()
+
+        m = client.scatter(model, broadcast=False)
+        subsample = 8
+
+        def apply_mast3r(model, image1: Image, image2: Image) -> Tuple[np.ndarray, np.ndarray, np.ndarray, np.ndarray]:
+            """Applies the MASt3R model to a pair of images to generate correspondences.
+
+            Args:
+                model: The MASt3R model instance.
+                image1: The first image.
+                image2: The second image.
+
+            Returns:
+                A tuple containing:
+                - matches_im1: Array of 2D coordinates of correspondences in the first image.
+                - matches_im2: Array of 2D coordinates of correspondences in the second image.
+                - idx1: Indices of correspondences in the first image's original feature grid.
+                - idx2: Indices of correspondences in the second image's original feature grid.
+            """
+            image1_proc, crop1, scale1 = preprocess_image(image1, 0, device)
+            image2_proc, crop2, scale2 = preprocess_image(image2, 1, device)
+
+            with torch.no_grad():
+                model = model.to(device)
+
+                res = symmetric_inference(model, image1_proc, image2_proc, device=device)
+                X11, X21, X22, X12 = [r["pts3d"][0] for r in res]
+                C11, C21, C22, C12 = [r["conf"][0] for r in res]
+                descs = [r["desc"][0] for r in res]
+                qonfs = [r["desc_conf"][0] for r in res]
+
+                idx1, idx2, matches_im1, matches_im2, scores = extract_correspondences(
+                    descs, qonfs, device=device, subsample=subsample
+                )
+                conf_score = (C11.mean() * C12.mean() * C21.mean() * C22.mean()).sqrt().sqrt()
+                # TODO: use the matching score to filter correspondences.
+                matching_score = (float(conf_score), float(scores.sum()), len(scores))
+
+                matches_im1 = ((matches_im1 + np.array(crop1)) + 0.5) / scale1
+                matches_im2 = ((matches_im2 + np.array(crop2)) + 0.5) / scale2
+
+            return (matches_im1, matches_im2, idx1, idx2)
```

Comment:
I'm a little worried about the readability of this function with two nested functions

Reply:
fair, I moved one of them outside. The other is designed to be a nested function (it changes those dicts inside it). I can force it to be non-nested if you prefer, but I dont think it hurts readability much as is.